scrapers/StateOfTexas.mjs

```diff
@@ -1,32 +1,67 @@
 import fetch from 'node-fetch';
 import moment from 'moment';
 
+const API = {
+  baseUrl: 'https://api.tnris.org/api/v1',
+
+  async getCollectionsCatalog(url) {
+    return await (await fetch(url)).json();
+  },
+  async getCollection(collectionId) {
+    const response = await fetch(`${this.baseUrl}/collections/${collectionId}`);
+    if (!response.ok)
+      return Promise.reject(
+        `Server responded with status ${response.status} for collection ${collectionId}`
+      );
+    return response.json();
+  },
+  async getResources(collectionId) {
+    let nextUrl = `${this.baseUrl}/resources?collection_id=${collectionId}`;
+    let results = [];
+
+    while (nextUrl !== null) {
+      const resources = await (await fetch(nextUrl)).json();
+      results.push(...resources.results);
+      nextUrl = resources.next;
+    }
+    return results;
+  },
+};
+
 export default ({ database, DataScraper }) => {
-	return DataScraper(database, 'StateOfTexas', async () => {
-		const data = [];
-		const baseUrl = 'https://api.tnris.org/api/v1';
-		const id = '2679b514-bb7b-409f-97f3-ee3879f34448';
-		const collections = `collections/${id}`;
-		const resources = `resources?collection_id=${id}`;
-		const collectionsRequest = await fetch(`${baseUrl}/${collections}`);
-		const collection = await collectionsRequest.json();
-		const resourcesRequest = await fetch(`${baseUrl}/${resources}`);
-		const resource = await resourcesRequest.json();
-
-		const { publication_date, description, name } = collection;
-		const { results } = resource;
-		const created = moment(publication_date, 'YYYY-MM-DD').unix();
-
-		results.map(item => {
-			return data.push({
-				url: item.resource,
-				updated: 0,
-				created,
-				description,
-				name,
-			});
-		});
-
-		return data;
-	});
+  return DataScraper(database, 'StateOfTexas', async () => {
+    const data = [];
+
+    //limit to 100 results per request to avoid overloading server
+    let nextCatalogUrl = `${API.baseUrl}/collections_catalog?limit=100`;
+
+    while (nextCatalogUrl !== null) {
+      console.log('scraping datasets for endpoint:', nextCatalogUrl);
+      const { results: catalogResults, next } = await API.getCollectionsCatalog(nextCatalogUrl);
+
+      // create array of promises that resolve concurrently without waiting for each other
+      const promises = catalogResults.map(result => {
+        return Promise.all([
+          API.getCollection(result.collection_id),
+          API.getResources(result.collection_id),
+        ]).then(([collection, resources]) => {
+          for (let item of resources) {
+            data.push({
+              url: item.resource,
+              updated: 0,
+              created: moment(collection.publication_date, 'YYYY-MM-DD').unix(),
+              description: collection.description,
+              name: collection.name,
+            });
+          }
+        });
+      });
+
+      await Promise.allSettled(promises);
```

Comment:
Using `Promise.all` or friends in this manner isn't a great idea. It will spam requests to the API and we might get rate limited, I would want to avoid this headache.

Reply:
Yeah, I noticed I was getting rate limited when I didn't restrict it to a limit, so I decided to do 100 at a time and that worked fine. I could've also sent requests one at a time but it would be slower as each promise has to wait until the previous has finished. So if speed isn't an issue than that can be a better solution.